src/store/CartProvider.js

```diff
@@ -7,31 +7,43 @@ const defaultCartState = {
 };
 
 const cartReducer = (state, action) => {
-   switch (action.type) {
-      case 'ADD':
-         const updatedTotalAmount = state.totalAmount + action.item.price * action.item.amount;
-         
-         const existingItemIndex = state.items.findIndex(item => item.id === action.item.id);
-         const existingCartItem = state.items[existingItemIndex];
-
-         let updatedItem;
-         let updatedItems;
-
-         if (existingCartItem) {
-            updatedItem = {
-               ...existingCartItem,
-               amount: existingCartItem.amount + action.item.amount
-            }
-            updatedItems = [...state.items];
-            updatedItems[existingItemIndex] = updatedItem;
+   if (action.type === 'ADD') {
```

Comment:
I think using the switch case statement here is OK.
can consider extracting those calculations to utils functions even in the same file. It adds better abstraction layers to what happened here. 
To calculate based on previous amount value seems a bit dangerous. I would prefer to calculate based on all items in the current state all over again.

```es6
const cartReducer = (state, action) => {
  let updatedItems
  switch (action.type) {
    case 'ADD':
      updatedItems = _placeItemIntoItems({
        items: state.items,
        changedItem: action.item
      })
      const updatedTotalAmount = _calculateTotal(updatedItems)
      return {items: updatedItems, totalAmount: updatedTotalAmount};
    case 'REMOVE':
      // remove
    default:
      // default
  }
}


function _calculateTotal(items){
  return items.reduce((total, item) => total + item.price * item.amount, 0);
}
function _placeItemIntoItems({items, changedItem}){
  let tempItems = [...items];
  const existingItemIndex = tempItems.findIndex(item => item.id === changedItem.id)
  if (existingItemIndex === -1) {
    return tempItems.concat(action.item);
  }
  const existingCartItem = tempItems[existingItemIndex]
  const updatedItem = {
    ...existingCartItem,
    amount: existingCartItem.amount + changedItem.amount
  }
  tempItems[existingItemIndex] = updatedItem;
  return tempItems
}
```

Reply:
Thanks for the suggestions! I changed from `switch` to `if` statements because of minor variable naming issues. 

You said: _"To calculate based on previous amount value seems a bit dangerous."_  Why is it dangerous? Is it because it might not be the correct current value? But I thought the `state` in useReducer always has the latest snapshot...